File: Python/6.ZigZag Conversion.py
#字符串以Z字形摆放，按照每横行的顺序重新输出
class Solution(object):
    def convert(self, s, numRows):
        ls = len(s)
        if  numRows==1 or numRows >= ls or ls <= 1:
            return s
        L = ""
        for i in range(0,numRows):
            j = i
            down = (numRows-i)*2-2 if i!=numRows-1 else 0
            up = i*2 if i!=0 else 0
            L += s[j]
            j = j+down if down!=0 else j+up
            while j<ls and len(L)<ls:
                if up!= 0:
                    L += s[j]
                    j += up
                if len(L)>=ls:
                    break
                if down!=0 and j<ls:
                    L += s[j]
                    j += down


        return L

s = Solution()
l = s.convert("ABCDE",4)
print(l)

File: C++/String/6.ZigZag Conversion.cpp
/*
字符串以Z字形摆放，按照每横行的顺序重新输出.
"PAYPALISHIRING"摆放为
P   A   H   N
A P L S I I G
Y   I   R
返回"PAHNAPLSIIGYIR"。
*/
/*
思路：第一行和最后一行的间隔为2*(numRows-1)
	中间第i行的间隔为：2*(numRows-1)-2*i和i*2交替。
*/
#include <iostream>
#include <cstring>
#include <vector>
using namespace std;

class Solution {
public:
	string convert(string s, int numRows) {
		if(numRows==1)  return s;
        int gap = 2 * (numRows-1) , length = s.length();;
        string result = "";
        for(int j = 0; j < length; j += gap)
        	result += s[j];
        for(int i = 1; i < numRows-1; ++i)
        {
        	int newgap = 2*i;
        	for(int j=i; j < length; j += newgap)
        	{
        		result += s[j];
        		newgap = gap - newgap;
        	}
        }
        for(int j = numRows-1; j < length; j += gap)
        	result += s[j];
        return result;
    }
};

int main()
{
	Solution s;;
	cout<<s.convert("PAYPALISHIRING",3)<<endl;
	cout<<s.convert("ABCDEFGHIJKLMN",4)<<endl;
}

